Library spring data mongodb

kondury opened the pull request.

kondury committed:
hw 8 spring data + mongodb, initial commit

kondury committed:
hw 8 spring dat, mongodb, mongock, embedded mongodb for tests + spring shell
- migrated repositoties on MongoRepository
- migrated services
- added mongock migration sets
- added cascade deletion for comments when book is deleted
- added cascade removing for genres in books (now books can have no genre)
- added unit tests for services
- added tests for non-standard repository operations, cascading included
- added jline-jna-terminal for autocompletion support when app is started via 'java -jar ...'

stvort commented on library/library-spring-data-mongodb/src/main/java/com/github/kondury/library/domain/Author.java:
```diff
@@ -0,0 +1,26 @@
+@Document("authors")
+public class Author {
+
+    public static final String UNKNOWN_AUTHOR = "<UNKNOWN AUTHOR>";
```
Зачем автор управляет своим внешним видом на UI?

stvort commented on library/library-spring-data-mongodb/src/main/java/com/github/kondury/library/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private Optional<Book> save(String bookId, String title, String authorId, String genreId) {
+        Optional<Genre> genre = genreId != null ? genreRepository.findById(genreId) : Optional.empty();
```
```
var genre = Optional.ofNullable(genreId)
    .map(genreRepository::findById)
    .orElse(null);
```
?

stvort commented on library/library-spring-data-mongodb/src/main/java/com/github/kondury/library/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+        Optional<Genre> genre = genreId != null ? genreRepository.findById(genreId) : Optional.empty();
+        var author = authorRepository.findById(authorId);
+
+        if (author.isPresent()) {
+            var book = bookRepository.save(new Book(bookId, title, author.get(), genre.orElse(null)));
+            return Optional.of(book);
+        }
```
Нет автора, не сохраняем книгу? Возможно это все же исключительная ситуация

stvort commented on library/library-spring-data-mongodb/src/test/java/com/github/kondury/library/repository/BookRepositoryTest.java:
```diff
@@ -0,0 +1,55 @@
+
+        bookRepository.deleteById(bookToDelete.getId());
+
+        assertThat(commentRepository.findByBookId(bookToDelete.getId())).isEmpty();
+        assertThat(commentRepository.findAll())
+                .extracting(v -> v.getBook().getId())
+                .doesNotContain(bookToDelete.getId());
```
Для проверки результата лучше, по возможности, обойтись средствами `MongoTemplate`

kondury committed:
hw 8 spring data, mongodb, mongock, embedded mongodb for tests + spring shell; after review fixes
- replaced auxiliary repository calls with mongoTemplate
- moved null-value marker to converter classes
- added EntityDoesNotExist exception
- removed Optional from insert and update Service methods
- made insert and update methods in Book and Comment services to throw exception if current or child entity do not exist
- added tests on exceptions thrown by services when some entities do not exist
